<a href="https://colab.research.google.com/github/corebankingfiap-dev/core_banking/blob/main/04_02_Correlacao_Segmentada.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Matriz de Correlação vai nos mostrar um número de -1 a +1:

+1 (Verde escuro): Andam grudados. Se um sobe, o outro sobe. (Ruim para proteção).

0 (Branco): Não têm relação nenhuma.

-1 (Vermelho escuro): Andam em direções opostas. Se um cai, o outro sobe. (O "Santo Graal" da diversificação).

In [1]:
import pandas as pd
import numpy as np
import os
from google.colab import drive

# 1. Montar o Drive
if not os.path.exists('/content/drive'):
    drive.mount('/content/drive')

# 2. CONFIGURAÇÃO (Troque para "bancos" ou "commodities" quando precisar)
NOME_SEGMENTO = "fiis"

# Ajustado para ler da SILVER (onde estão os preços diários das suas fotos)
ARQUIVO = f"b3_{NOME_SEGMENTO}_limpo.csv"
base_path_silver = '/content/drive/MyDrive/DATA_LAKE/02_Silver/'
base_path_gold = '/content/drive/MyDrive/DATA_LAKE/03_Gold/'

full_path = os.path.join(base_path_silver, ARQUIVO)

if os.path.exists(full_path):
    print(f"✅ Calculando correlação para: {NOME_SEGMENTO.upper()}")

    # Lendo os preços e calculando retornos (correlação se faz sobre retornos)
    df_precos = pd.read_csv(full_path, index_col=0, parse_dates=True)
    df_retornos = df_precos.pct_change().dropna()

    # 4. CÁLCULO DA MATRIZ DE CORRELAÇÃO (Pearson)
    matriz_corr = df_retornos.corr()

    # 5. FORMATAÇÃO PARA EXIBIÇÃO
    # Correlação vai de -1 a 1. Arredondamos para 2 casas.
    matriz_view = matriz_corr.round(2)

    # 6. SALVAMENTO (Dentro da Gold > REPORTS conforme seu padrão)
    path_reports = os.path.join(base_path_gold, 'REPORTS')
    os.makedirs(path_reports, exist_ok=True)

    # Salvando os arquivos
    matriz_corr.to_csv(os.path.join(path_reports, f"CORREL_{NOME_SEGMENTO.upper()}.csv"))

    print("\n--- MATRIZ DE CORRELAÇÃO (Diária) ---")
    print(matriz_view)
    print("-" * 50)
    print(f"Arquivo salvo em: {path_reports}/CORREL_{NOME_SEGMENTO.upper()}.csv")

else:
    print(f"❌ Arquivo não encontrado na Silver: {full_path}")

Mounted at /content/drive
✅ Calculando correlação para: FIIS

--- MATRIZ DE CORRELAÇÃO (Diária) ---
        HGLG11  HGRU11  KNRI11  XPLG11  XPML11
HGLG11    1.00    0.33    0.27    0.33   -0.00
HGRU11    0.33    1.00    0.26    0.30   -0.01
KNRI11    0.27    0.26    1.00    0.31   -0.02
XPLG11    0.33    0.30    0.31    1.00   -0.00
XPML11   -0.00   -0.01   -0.02   -0.00    1.00
--------------------------------------------------
Arquivo salvo em: /content/drive/MyDrive/DATA_LAKE/03_Gold/REPORTS/CORREL_FIIS.csv


Quando você rodar, verá uma tabela onde os ativos se cruzam.

Diagonal de 1.00: Você verá uma linha diagonal de 1.00. Isso é normal, é a correlação do ativo com ele mesmo.

Próximo de 1.00 (Ex: 0.85): Indica que os ativos "andam juntos". Se o Itaú sobe, o Bradesco provavelmente sobe.

Próximo de 0.00 (Ex: 0.05): Indica que os ativos são independentes. O movimento de um não explica o do outro.

Negativo (Ex: -0.30): O "Ouro". Significa que eles tendem a ir para lados opostos. Se um cai, o outro ajuda a segurar a carteira.